clients/drcachesim/tests/allasm_scattergather_aarch64.asm

```diff
@@ -0,0 +1,485 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This is a statically-linked app.
+ * XXX i#3107: Pure-asm apps like this are difficult to maintain; We need this to be
+ * pure-asm so that we can verify exact instruction, load and store counts for the
+ * expanded scatter/gather sequence. With support for #3107, we can use annotations to
+ * mark app phases and compute those counts for only the intended parts in the C+asm app.
+ *
+ * The drx-scattergather-aarch64.cpp test app covers functional testing of the expansion.
+ * This test just needs to execute some scatter/gather/predicated contiguous loads and
+ * stores so we can verify the counts in the basic_counts output are correct.
+ */
+.text
+        .global  _start
+
+        .align   6
+
+// Destination registers used in load instructions
+#define DEST_REG1 z28
+#define DEST_REG2 z29
+#define DEST_REG3 z30
+#define DEST_REG4 z31
+
+// Source registers used in store instructions. These alias DEST_REG* because this test
+// doesn't care about the actual values being loaded/stored.
+#define SRC_REG1 z28
+#define SRC_REG2 z29
+#define SRC_REG3 z30
+#define SRC_REG4 z31
+
+#define B_MASK_REG p0 // Governing predicate for byte-element instructions
+#define H_MASK_REG p1 // Governing predicate for half-element instructions
+#define S_MASK_REG p2 // Governing predicate for word-element instructions
+#define D_MASK_REG p3 // Governing predicate for doubleword-element instructions
+
+#define BUFFER_REG x1
+#define Z_BASE_REG z0 // base reg used in vector+immediate instructions
+
+#define S_INDEX_REG z1 // index reg used in scalar+immed instructions with 32-bit elements
+#define D_INDEX_REG z2 // index reg used in scalar+immed instructions with 64-bit elements
+
+#define X_INDEX_REG x2 // index reg used in scalar+scalar instructions
+
+/*
+ * Test functions. The commented number after each instruction indicates the number of
+ * elements this instruction accesses with a 128-bit vector length. We can add these
+ * numbers up to determine how many loads/stores we expect to see in the basic_counts
+ * output when all elements are active. To find the number for hardware with a larger
+ * vector size, multiply by vl_size/16.
+ */
+
+test_scalar_plus_vector:
+        // ld1b scalar+vector
+        ld1b    DEST_REG1.s, S_MASK_REG/z, [BUFFER_REG, S_INDEX_REG.s, uxtw] // 4
```

Comment:
This test doesn't verify the result of these operations under drmemtrace. Could we also run a test which enables drmemtrace on suite/tests/client-interface/drx-scattergather-aarch64.cpp so that we verify that drmemtrace doesn't corrupt app state (missing drreg restore etc)?

Reply:
We already have a test `tool.drcachesim.scattergather-aarch64` which runs `suite/tests/client-interface/drx-scattergather-aarch64.cpp` under `drcachesim`. Is that what you mean?

Reply:
Ah yes. Apologies. I somehow assumed that was just plain DR.